pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java

```diff
@@ -170,6 +171,30 @@ public List<String> getRestartableStages() {
         return stages;
     }
 
+    /**
+     * Returns whether a stage is restartable.
+     * @param stageName the stage name
+     * @return true is restartable, false otherwise
+     */
+    public boolean isRestartable(String stageName) {
+        FlowExecution execution = getExecution();
+        if (execution != null) {
+            ExecutionModelAction execAction = run.getAction(ExecutionModelAction.class);
+            if (execAction != null) {
+                ModelASTStages stages = execAction.getStages();
+                if (stages != null) {
+                    for (ModelASTStage s : stages.getStages()) {
+                        if(s.getName().equals(stageName)) {
+                            return !Utils.stageHasStatusOf(s.getName(), execution,
+                                StageStatus.getSkippedForFailure(), StageStatus.getSkippedForUnstable());
```

Comment:
So a failed/unstable stage is _not_ restartable? The logic here is unclear. Is this just a copy / adaptation of code elsewhere?

Reply:
@jglick it is the same logic as [isRestartable()](https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/a3f80da24c5bc4ad366c60b86165b845df4c2327/pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java#L154-L172) but specific to a single stage.